### PR TITLE
Fixed #35257 -- Corrected resolving output_field for IntegerField/DecimalField with NULL.

### PR DESCRIPTION
**Pull request by sharonwoo**
https://code.djangoproject.com/ticket/35257

### CODE REVIEW

**Comment by felixxm** [comment]
Please add a regression test.

File: tests/expressions/tests.py [hunk 1]
```diff
@@ -2333,6 +2333,34 @@ def test_resolve_output_field(self):
                 expr = Value(value)
                 self.assertIsInstance(expr.output_field, output_field_type)
 
+    def test_mixed_output_field_with_none_returns_field(self):
```
Comment (sharonwoo) [inline tests/expressions/tests.py:2336]: This will probably not need to be covered with https://github.com/django/django/pull/17890, now.

File: tests/expressions/tests.py [hunk 2]
```diff
@@ -2333,6 +2333,34 @@ def test_resolve_output_field(self):
...
+            Value(datetime.time(3, 16)) + Value(None),
+            Value(datetime.datetime(2019, 5, 15)) + Value(None),
+            Value(datetime.timedelta(1)) + Value(None),
+            Value(b"") + Value(None),
+            Value(uuid.uuid4()) + Value(None),
+        ]
+
+        for expression in test_cases:
+            with self.subTest(expression):
+                with self.assertRaises(FieldError):
```
Comment (sharonwoo) [inline tests/expressions/tests.py:2361]: To update the FieldError to `OutputFieldIsNoneError` in https://github.com/django/django/pull/17890

Comment (sharonwoo) [reply tests/expressions/tests.py:2361]: @charettes if I understand you correctly...

1. If we infer the output field (e.g. the nonsensical test cases here to trigger inference of `output_field` through `_connector_combinations`), we **_should not_** raise? 
This would be different from existing behaviour which was established here: https://github.com/django/django/pull/15271

2. If explicitly specifying the output field e.g. with `ExpressionWrapper(Value(...), Value(...), output_field=DateTimeField())`, similar to `Value(None, output_field=DateTimeField(null=True))`, this does not raise.

File: tests/expressions/tests.py [hunk 3]
```diff
@@ -2333,6 +2333,34 @@ def test_resolve_output_field(self):
...
+        test_cases = [
+            Value(None),
+            Value("a string") + Value(None),
+            Value(datetime.date(2019, 5, 15)) + Value(None),
+            Value(datetime.time(3, 16)) + Value(None),
+            Value(datetime.datetime(2019, 5, 15)) + Value(None),
+            Value(datetime.timedelta(1)) + Value(None),
+            Value(b"") + Value(None),
+            Value(uuid.uuid4()) + Value(None),
+        ]
```
Comment (felixxm) [inline tests/expressions/tests.py:2357]: As far as I'm aware these test are unrelated with a patch so we should proceed them separately. Please remove them them from this PR.

File: tests/expressions/tests.py [hunk 4]
```diff
@@ -2333,6 +2333,34 @@ def test_resolve_output_field(self):
                 expr = Value(value)
                 self.assertIsInstance(expr.output_field, output_field_type)
 
+    def test_mixed_output_field_with_none_returns_field(self):
+        test_cases = [
+            (Value(3.14159) + Value(None), FloatField),
+            (Value(42) + Value(None), IntegerField),
+            (Value(Decimal("3.14")) + Value(None), DecimalField),
```
Comment (felixxm) [inline tests/expressions/tests.py:2340]: Please add tests for `None` on the left-hand side:
```suggestion
            (Value(Decimal("3.14")) + Value(None), DecimalField),
            (Value(None) + Value(3.14159), FloatField),
            (Value(None) + Value(42), IntegerField),
            (Value(None) + Value(Decimal("3.14")), DecimalField),
```

Comment (sharonwoo) [reply tests/expressions/tests.py:2340]: Sorry I was unwell for a bit, coming back to this now, let me work on it.

File: tests/expressions/tests.py (hunk 3 above)
Comment (sharonwoo) [reply tests/expressions/tests.py]: Removed~

File: tests/expressions/tests.py (hunk 4 above)
Comment (sharonwoo) [reply tests/expressions/tests.py:2340]: Added.

Comment (sharonwoo) [reply tests/expressions/tests.py:2340]: Exhaustive version of this test for comparison: 

```
    def test_mixed_output_field_with_none_returns_field(self):
        def generate_test_cases():
            test_values_and_fields = [
                (3.14159, FloatField),
                (42, IntegerField),
                (Decimal("3.14"), DecimalField),
            ]
            connectors = [
                Combinable.ADD,
                Combinable.SUB,
                Combinable.MUL,
                Combinable.DIV,
                Combinable.MOD,
                Combinable.POW
            ]
            test_cases = []

            for value, result_field in test_values_and_fields:
                for connector in connectors:
                    test_cases.append((Value(value)._combine(Value(None),
                                                             connector,
                                                             reversed=False),
                                       result_field))
                    test_cases.append((Value(value)._combine(Value(None),
                                                             connector,
                                                             reversed=True),
                                       result_field))

            return test_cases

        test_cases = generate_test_cases()

        for expression, result in test_cases:
            with self.subTest(expression):
                self.assertIsInstance(expression.output_field, result)
```

File: tests/expressions/tests.py (hunk 2 above)
Comment (sharonwoo) [reply tests/expressions/tests.py]: Note to self: 
Need to understand why in `test_resolve_output_field_with_null`, `(IntegerField, Combinable.ADD, null),` for example will trigger the error message `Cannot infer type...` when it should be able to infer through `_connector_combinations`

**Comment by felixxm** [comment]
@sharonwoo Thanks :+1: Welcome aboard :boat: 

I simplified tests.

**Comment by shangxiao** [comment]
Woo! First PR accepted 👏
